File: day_1/solution.java
import java.nio.file.Files;
import java.nio.file.Path;

class solution {
    private interface DigitFinder {
        int getDigit(String s, int idx);
    }

    private static int parseDigits(DigitFinder df, String s) {
        int a = -1, b = -1;
        for (int i = 0; i < s.length(); i++) {
            int an = df.getDigit(s, i);
            int bn = df.getDigit(s, s.length() - i - 1);
            if (a == -1 && an != -1) a = an;
            if (b == -1 && bn != -1) b = bn;
            if (a != -1 && b != -1) break;
        }
        return a * 10 + b;
    }

    private static class Part1DigitFinder implements DigitFinder {
        public int getDigit(String s, int idx) {
            if (Character.isDigit(s.charAt(idx))) {
                return s.charAt(idx) - '0';
            }
            return -1;
        }
    }

    private static void part1() throws Exception {
        var df = new Part1DigitFinder();
        int sum = Files.readAllLines(Path.of("part_1_input.txt")).stream()
            .mapToInt(line -> parseDigits(df, line))
            .sum();
        System.out.println(sum);
    }

    private static class Part2DigitFinder implements DigitFinder {
        public int getDigit(String s, int idx) {
            if (Character.isDigit(s.charAt(idx))) return s.charAt(idx) - '0';
            String[] words = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
            for (int i = 0; i < words.length; i++) {
                String word = words[i];
                if (s.length() - idx >= word.length() && s.substring(idx, idx + word.length()).equals(word)) {
                    return i + 1;
                }
            }
            return -1;
        }
    }

    private static void part2() throws Exception {
        var df = new Part2DigitFinder();
        int sum = Files.readAllLines(Path.of("part_2_input.txt")).stream()
            .mapToInt(line -> parseDigits(df, line))
            .sum();
        System.out.println(sum);
    }

    public static void main(String[] args) throws Exception {
        part1();
        part2();
    }
}

File: day_1/solution.lua
local function parseDigits(f, s)
    local a = -1
    local b = -1
    for i = 1, #s do
        local an = f(s, i)
        local bn = f(s, #s - i + 1)
        if a == -1 and an ~= -1 then a = an end
        if b == -1 and bn ~= -1 then b = bn end
        if a ~= -1 and b ~= -1 then break end
    end
    return a * 10 + b
end

local function part1()
    local f = io.open("part_1_input.txt")
    local sum = 0
    local line = f:read("*line")
    while line ~= nil do
        sum = sum + parseDigits(
            function (s, i)
                local num = tonumber(string.sub(s, i, i))
                if num ~= nil then return num else return -1 end
            end,
            line
        )
        line = f:read("*line")
    end
    f:close()
    print(sum)
end

local function part2()
    local function findDigit(s, idx)
        local num = tonumber(string.sub(s, idx, idx))
        if num ~= nil then return num end
        local words = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
        for i = 1, #words do
            local word = words[i]
            if #s - idx + 1 >= #word and string.sub(s, idx, idx + #word - 1) == word then
                return i
            end
        end
        return -1
    end
    local f = io.open("part_2_input.txt")
    local sum = 0
    local line = f:read("*line")
    while line ~= nil do
        sum = sum + parseDigits(findDigit, line)
        line = f:read("*line")
    end
    print(sum)
end

part1()
part2()
